From bc22a48a46bd61ea52985cfc2c92df5c4a78435a Mon Sep 17 00:00:00 2001
From: Prarit Bhargava <prarit@redhat.com>
Date: Thu, 28 Aug 2014 14:22:06 -0400
Subject: [cpufreq] don't allow governor limits to be changed when it is disabled

Message-id: <1409235879-1283-33-git-send-email-prarit@redhat.com>
Patchwork-id: 89403
O-Subject: [RHEL7.1 PATCH BZ 1134639 032/185] cpufreq: don't allow governor limits to be changed when it is disabled
Bugzilla: 1134369
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Mark Langsdorf <mlangsdo@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1134639

commit f73d39338444d9915c746403bd98b145ff9d2ba4
Author: Viresh Kumar <viresh.kumar@linaro.org>
Date:   Sat Aug 31 17:53:40 2013 +0530

    cpufreq: don't allow governor limits to be changed when it is disabled

    __cpufreq_governor() returns with -EBUSY when governor is already
    stopped and we try to stop it again, but when it is stopped we must
    not allow calls to CPUFREQ_GOV_LIMITS event as well.

    This patch adds this check in __cpufreq_governor().

    Reported-by: Stephen Boyd <sboyd@codeaurora.org>
    Signed-off-by: Viresh Kumar <viresh.kumar@linaro.org>
    Signed-off-by: Rafael J. Wysocki <rafael.j.wysocki@intel.com>

Cc: Lenny Szubowicz <lszubowi@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/cpufreq/cpufreq.c b/drivers/cpufreq/cpufreq.c
index 239a95f..7cd53d0 100644
--- a/drivers/cpufreq/cpufreq.c
+++ b/drivers/cpufreq/cpufreq.c
@@ -1699,8 +1699,9 @@ static int __cpufreq_governor(struct cpufreq_policy *policy,
       policy->cpu, event);
 
  mutex_lock(&cpufreq_governor_lock);
- if ((!policy->governor_enabled && (event == CPUFREQ_GOV_STOP)) ||
-     (policy->governor_enabled && (event == CPUFREQ_GOV_START))) {
+ if ((policy->governor_enabled && event == CPUFREQ_GOV_START)
+     || (!policy->governor_enabled
+     && (event == CPUFREQ_GOV_LIMITS || event == CPUFREQ_GOV_STOP))) {
   mutex_unlock(&cpufreq_governor_lock);
   return -EBUSY;
  }
-- 
1.7.1